v2/internal/util/metrictree/tree.go

```diff
@@ -115,49 +131,89 @@ func (tv *TreeValidator) AddRule(ns string) error {
 		}
 	case metricFilteringStrategy:
-		defPresent := tv.definitionTree.HasRules()
-		if !parsedNs.IsUsableForFiltering(defPresent) {
+		defPresent := tv.definitionTree.hasRules()
+		if !parsedNs.IsUsableForFiltering(tv.constraints, defPresent) {
 			return fmt.Errorf("can't add rule (%s) - some namespace elements are not allowed in filtering when metric definition wasn't provided", ns)
 		}
 
-		if !tv.definitionTree.IsCompatible(ns) {
+		if !tv.definitionTree.isCompatible(ns) {
 			return fmt.Errorf("can't add rule (%s) - not compatible with any metric definition", ns)
 		}
 	default:
 		panic("invalid strategy")
 	}
 
-	return tv.updateTree(parsedNs)
+	return tv.updateTree(parsedNs, tv.constraints)
 }
 
+func (tv *TreeValidator) IsUsableForAddition(ns string, isFilter bool) error {
+	parsedNs, err := ParseNamespace(ns, false)
+	if err != nil {
+		return fmt.Errorf("invalid format of namespace: %v", err)
+	}
+
+	ok := parsedNs.IsUsableForAddition(tv.constraints, tv.hasRules(), isFilter)
+	if !ok {
+		return errors.New("metric not usable for addition")
+	}
+
+	return nil
+}
+
+// IsPartiallyValid does a partial metric validation in metricFilteringStrategy
+// is used by ctx.ShouldProcess to provide quick-return optimization in collecting metrics routine(s)
 func (tv *TreeValidator) IsPartiallyValid(ns string) bool {
-	isValid, _ := tv.isValid(ns, false, false)
+	isValid, _ := tv.isValid(ns, false)
 	return isValid
 }
 
+// IsValid does full metric validation in metricFilteringStrategy
+// tests metric eligibility for adding, is called by ctx.AddMetric
 func (tv *TreeValidator) IsValid(ns string) (bool, []string) {
-	isValid, trace := tv.isValid(ns, true, false)
+	isValid, trace := tv.isValid(ns, true)
 	return isValid, trace
 }
 
-func (tv *TreeValidator) IsCompatible(ns string) bool {
-	isCompatible, _ := tv.isValid(ns, false, true)
+// isCompatible does partial metric validation in metricDefinitionStrategy
+// is used by AddRule to assure the new metric definition does not break out of the existing tree structure
+func (tv *TreeValidator) isCompatible(ns string) bool {
+	isCompatible, _ := tv.isValid(ns, false)
 	return isCompatible
 }
 
-func (tv *TreeValidator) HasRules() bool {
+func (tv *TreeValidator) hasRules() bool {
 	return tv.head != nil
 }
 
-func (tv *TreeValidator) isValid(ns string, fullMatch bool, compatibilityMode bool) (bool, []string) {
-	if compatibilityMode && tv.strategy != metricDefinitionStrategy {
-		panic("compatibilityMode can be only used for definition tree")
-	}
+func (tv *TreeValidator) AllowDynamicLastElement() {
+	tv.constraints &= ^lastNamespaceElementMustBeStatic
+}
 
+func (tv *TreeValidator) AllowValuesAtAnyNamespaceLevel() {
+	tv.constraints &= ^onlyLeavesCanHoldValues
+}
+
+func (tv *TreeValidator) AllowAddingUndefinedMetrics() {
+	tv.constraints &= ^rejectUndefinedNamespaces
+}
+
+func (tc TreeConstraints) lastNamespaceElementMustBeStatic() bool {
```

Comment:
(tv *TreeValidator)

Reply:
disagree - it's intentionally taken a level lower and kept private

Reply:
https://golang.org/doc/faq#methods_on_values_or_pointers recommends consistency is such case.

Reply:
yeah, I know that rule but this one here is not against that

please note those are methods defined on another type

Reply:
ohh ... I see, could we move the definitions below TreeValidator definitions then? (just not mix defs of the diffent types)

Reply:
done